tests/reporting/test_reporting.py

```diff
@@ -1,108 +1,442 @@
-    firstRegistrationToken = contractsFixture.applySignature('RegistrationToken', reportingWindow.getRegistrationToken())
+    reportingWindow = sessionFixture.applySignature('ReportingWindow', market.getReportingWindow())
+    firstRegistrationToken = sessionFixture.applySignature('RegistrationToken', reportingWindow.getRegistrationToken())
+    reportingTokenNo = sessionFixture.getReportingToken(market, [2,0])
+    reportingTokenYes = sessionFixture.getReportingToken(market, [0,2])
+
+    # Tester 0, 1, 2, and 3 will buy registration tokens so they can report on the market
     firstRegistrationToken.register(sender=tester.k0)
     assert firstRegistrationToken.balanceOf(tester.a0) == 1
-    assert reputationToken.balanceOf(tester.a0) == 6 * 10**6 * 10**18 - 10**18
-    firstRegistrationToken.register(sender=tester.k1)
-    assert firstRegistrationToken.balanceOf(tester.a1) == 1
-    assert reputationToken.balanceOf(tester.a1) == 4 * 10**6 * 10**18 - 10**18
-
-    # fast forward to one second after the next reporting window
-    reportingWindow = contractsFixture.applySignature('ReportingWindow', branch.getNextReportingWindow())
-    contractsFixture.chain.head_state.timestamp = reportingWindow.getStartTime() + 1
-
-    # both reporters report on the outcome
-    reportingTokenNo = contractsFixture.getReportingToken(market, [2,0])
-    reportingTokenYes = contractsFixture.getReportingToken(market, [0,2])
+    # Testers will have their previous REP balance less the registration token bond of 1 REP
+    assert reputationToken.balanceOf(tester.a0) == 8 * 10**6 * 10**18 - 10**18
+
+    for (key, address) in [(tester.k1, tester.a1), (tester.k2, tester.a2), (tester.k3, tester.a3)]:
+        firstRegistrationToken.register(sender=key)
+        assert firstRegistrationToken.balanceOf(address) == 1
+        assert reputationToken.balanceOf(address) == 1 * 10**6 * 10**18 - 10**18
+
+    sessionFixture.reputationToken = reputationToken
+    sessionFixture.reportingWindow = reportingWindow
+    sessionFixture.firstRegistrationToken = firstRegistrationToken
+    sessionFixture.reportingTokenNo = reportingTokenNo
+    sessionFixture.reportingTokenYes = reportingTokenYes
+    return sessionFixture.chain.snapshot()
+
+@fixture
+def reportingFixture(sessionFixture, reportingSnapshot):
+    sessionFixture.chain.revert(reportingSnapshot)
+    return sessionFixture
+
+def test_reportingFullHappyPath(reportingFixture):
+    cash = reportingFixture.cash
+    market = reportingFixture.binaryMarket
+    reputationToken = reportingFixture.reputationToken
+    branch = reportingFixture.branch
+    reportingTokenNo = reportingFixture.reportingTokenNo
+    reportingTokenYes = reportingFixture.reportingTokenYes
+
+    # We can't yet report on the market as it's in the pre reporting phase
+    assert market.getReportingState() == reportingFixture.constants.PRE_REPORTING()
+    with raises(TransactionFailed, message="Reporting cannot be done in the PRE REPORTING state"):
```

Comment:
What does the `message` parameter here do?

Reply:
Yup!